src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -633,7 +686,11 @@ private void NotifyListeners(IProjectSnapshot older, IProjectSnapshot newer, str
     // virtual so it can be overridden in tests
     protected virtual void NotifyListeners(ProjectChangeEventArgs e)
     {
-        _projectSnapshotManagerDispatcher.AssertDispatcherThread();
+        _locksFactory.EnsureNoWriteLock();
+
+        // Get a read lock to make sure nothing changes while notifications
+        // are going out.
+        using var _ = _locksFactory.EnterReadLock();
```

Comment:
lol, I guess this is a direct counterpoint to the "are we sure we want the notifications to be done under a lock".

Reply:
Right now, what happens if a receiver calls into some code that tries to take a lock, of any kind? Does it throw? Does it `Debug.Assert` (assuming the assert in the fault handler is put back)?

I guess my real question is: How do we know that we don't have any receivers that violate this now?

Reply:
We don't, but one of two things would happen: 

1. They are attempting something that requires a read lock, which would succeed just fine and have the correct current state of the world as the notification.
2. They try to get a write lock, which in release builds will timeout after 30 seconds and throw an exception.

Reply:
This seems pretty scary to me. Am I wrong?

Previously, notifications would happen on the dispatcher thread, so any write operations would just about be guaranteed to succeed, or the operation could switch to the dispatcher thread. With this change, write operations are guaranteed to time out and throw an exception.

Reply:
Yea, I'm looking at usage more and this will probably break something. I'll remove the need for a read lock for now. What I do worry about is that before something like the following would always be true, and now it isn't. 

```c#
void OnChanged(object sender, ProjectChangeEventArgs e) 
{
    Assert.Equal(e.NewProject, _snapshotManager.GetLoadedProject(e.FilePath));
}
```